ports/stm32/powerctrl.c

```diff
@@ -116,9 +117,14 @@ NORETURN static __attribute__((naked)) void branch_to_bootloader(uint32_t r0, ui
         );
     MP_UNREACHABLE;
 }
+#endif
 
 NORETURN void powerctrl_enter_bootloader(uint32_t r0, uint32_t bl_addr) {
-    #if MICROPY_HW_ENTER_BOOTLOADER_VIA_RESET
+    #if defined(MICROPY_RESET_TO_BOOTLOADER)
+
+    MICROPY_RESET_TO_BOOTLOADER();
```

Comment:
Instead of this I think it would be cleaner for a board to be able to override the behaviour of ``()`.  Because then the logic of `MICROPY_HW_ENTER_BOOTLOADER_VIA_RESET` remains intact and can be used in conjunction with a custom implementation of `branch_to_bootloader()` (or not used if not needed).

I suggest to do it like other macros in `boardctrl.h`:
- rename `branch_to_bootloader` to `powerctrl_branch_to_bootloader`, and remove the "static" keyword from it
- add the declaration of that function to `powerctrl.h`
- add a default `#define MICROPY_BOARD_BRANCH_TO_BOOTLOADER powerctrl_branch_to_bootloader` in `boardctrl.h`

Reply:
Are you saying override `branch_to_bootloader` and do the RTC & `NVIC_SystemReset()` logic in there ? While not override `powerctrl_enter_bootloader` ?

Reply:
For example, in `boardctrl.h` or `powerctrl.h`:
```c
#ifndef MICROPY_BOARD_ENTER_BOOTLOADER
#define MICROPY_BOARD_ENTER_BOOTLOADER powerctrl_enter_bootloader
#endif
```

Then in `modmachine.c` call `MICROPY_BOARD_ENTER_BOOTLOADER(0x.., MBOOT_VTOR)` instead. I would also define the same macro in `nrf` port for Nano 33 BLE,  and call it from `modmachine.c` if defined, since there's no `power/boardctrl` in there.

Reply:
> If MICROPY_HW_ENTER_BOOTLOADER_VIA_RESET is disabled then powerctrl_enter_bootloader() is just calling branch_to_bootloader() and doing nothing else.

For now but what if this behavior needs to change in the future ? It may somehow break it, that's why I think it makes sense to allow overriding the whole function and not just the branching part.

> branch_to_bootloader() is the final function that actually transfers control to the bootloader

It's a technicality, but I don't think the macro name reflects what it's actually doing, since it's not entering the bootloader via a branch.

One final point, `MICROPY_BOARD_ENTER_BOOTLOADER` is more generic an can be used in other ports which don't have any code that enters bootloaders (ex the nrf) or ports in which entering a bootloader does not necessarily involve branching (ex rp2).

Reply:
The way stm32 gets into the bootloader is complicated and supports a few different ways, and supports different bootloaders.  And I don't want to make it more complicated with config options that aren't orthogonal to existing ones.

I suggest we open a new PR just for this bootloader issue, and continue discussion there.  I agree it would be good to make it consistent across ports.

Reply:
@dpgeorge Okay I removed the second commit, and there's no special bootloader config options now. This PR should be good to merge (with the image in media). Will open an issue about the bootloader and we can discuss further.

NOTE: upli.c was missing a header include.

Reply:
See  #8590